### PR TITLE
Set locales for all domains when enabling multilingual

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
In the process of trying to test a possible wordpress regression I found that enabling
mutlilingual on a domain other than domain 1 breaks the site. The reason is that without
only domain 1 is being updated with the new language and locales. Since I'm on domain
2 when I do this I can no longer load any page as the locales value has not been set
and the queries look for option_value.label not option_value.label_us.

Since 'no locale' is no longer valid the locale needs to be set on all sites. Note the
code change is primarily an extraction inn order to only run the table update part once

Before
----------------------------------------
When enabling multilinngual for one domain 'locales' not updated for other domains and they become unusable (wordpress)

After
----------------------------------------
civicrm_domain.locales set for all domains to the language being enabled

Technical Details
----------------------------------------
@colemanw you attempt to address this here - https://github.com/civicrm/civicrm-core/pull/17650/files#diff-fc3e5db5312bd5c376b4b73c3c1be43fR46 but in fact it needs to be all domains get locales set as it is set for the tables that all domains access

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Core/I18n/Schema.php [hunk 1]
```diff
@@ -44,56 +44,26 @@ public static function dropAllViews() {
...
-        $queries[] = "ALTER TABLE {$table} ADD {$column}_{$locale} {$type}";
-        if (CRM_Core_BAO_SchemaHandler::checkIfFieldExists($table, $column)) {
-          $queries[] = "UPDATE {$table} SET {$column}_{$locale} = {$column}";
-          $queries[] = "ALTER TABLE {$table} DROP {$column}";
-        }
+    $domain->find();
+    while ($domain->fetch()) {
+      // skip if the domain is already multi-lang.
+      if ($domain->locales) {
+        continue;
```
Comment (seamuslee001) [inline CRM/Core/I18n/Schema.php:53]: Should we make sure that all domains have the same locales?

Comment (eileenmcnaughton) [reply CRM/Core/I18n/Schema.php:53]: @seamuslee001 I dunno - I think one domain could have less than the others but it gets tricky fast. I decided not to change that edge case

Comment (seamuslee001) [reply CRM/Core/I18n/Schema.php:53]: @colemanw was thinking that instead maybe domain1 should be treated as  global domain see other discussion here https://github.com/civicrm/civicrm-core/pull/17650

File: CRM/Core/I18n/Schema.php [hunk 2]
```diff
@@ -44,56 +44,26 @@ public static function dropAllViews() {
...
-      }
-      // deal with columns
-      foreach ($hash as $column => $type) {
-        $queries[] = "ALTER TABLE {$table} ADD {$column}_{$locale} {$type}";
-        if (CRM_Core_BAO_SchemaHandler::checkIfFieldExists($table, $column)) {
-          $queries[] = "UPDATE {$table} SET {$column}_{$locale} = {$column}";
-          $queries[] = "ALTER TABLE {$table} DROP {$column}";
-        }
+    $domain->find();
+    while ($domain->fetch()) {
```
Comment (eileenmcnaughton) [inline CRM/Core/I18n/Schema.php:50]: Something is wonky here... I think some cache is being reset as only first is retrieved

Comment (eileenmcnaughton) [reply CRM/Core/I18n/Schema.php:51]: OK - switched to putting in an array first...

**Comment by colemanw** [comment]
So historically domain 1 was treated as the global domain, and I think I messed that up in #17646 and that caused the regression. We've always set the locales value in domain 1 (or technically the first domain since I guess you could delete domain id 1) and my PR changed that. We should probably change it back.

**Comment by eileenmcnaughton** [comment]
@colemanw I'm unclear as to whether what I hit was 'the regression' - since it was a bit edge-case but yeah maybe it did regress. I think this change probably still makes sense even if we do revert the other

**Comment by colemanw** [comment]
I don't want to completely revert #17646 because the performance boost is a big deal and that was the main focus. Reading from the current domain was secondary so I'll just remove that part.

**Comment by eileenmcnaughton** [comment]
Hmm tests pass but no test site :-(

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@seamuslee001 any idea about the lack of test site?

**Comment by seamuslee001** [comment]
@eileenmcnaughton test site seems to be up here http://core-17733-4dpkx.test-1.civicrm.org:8001/

**Comment by seamuslee001** [comment]
at least for the latest test

**Comment by colemanw** [comment]
@eileenmcnaughton @seamuslee001 @totten we have multiple discussion threads about this and multiple open PRs.
I've changed my mind about the solution and I think having `locales` populated consistently for all domains is a good thing.
https://lab.civicrm.org/dev/core/-/issues/1852

**Comment by colemanw** [comment]
This is missing an upgrade step so I added one: #17738

File: tests/phpunit/CRM/Core/I18n/SchemaTest.php [hunk 3]
```diff
@@ -40,9 +40,17 @@ public function tearDown() {
    * @param string $expectedRewrite
    *
    * @dataProvider translateTables
+   * @throws \CRM_Core_Exception
    */
   public function testI18nSchemaRewrite($table, $expectedRewrite) {
     CRM_Core_I18n_Schema::makeMultilingual('en_US');
+    $domains = $this->callAPISuccess('Domain', 'get')['values'];
+    foreach ($domains as $domain) {
```
Comment (colemanw) [inline tests/phpunit/CRM/Core/I18n/SchemaTest.php:48]: Let's make sure we are actually testing multiple domains:

```suggestion
    $domains = $this->callAPISuccess('Domain', 'get')['values'];
    $this->assertGreaterThan(1, count($domains));
    foreach ($domains as $domain) {
```

**Comment by eileenmcnaughton** [comment]
This was merged to 5.28 -as was https://github.com/civicrm/civicrm-core/pull/17646.

It fixed a temporary regression but the regression was not in any rc

**Pull request merged**